In [1]:
def fact(n):
    if n==1:
        return 1
    return fact(n-1)*n
fact(5)

120

如果我们计算fact(5)，可以根据函数定义看到计算过程如下：

 fact(5)
 - 5 * fact(4)
 - 5 * (4 * fact(3))
 - 5 * (4 * (3 * fact(2)))
 - 5 * (4 * (3 * (2 * fact(1))))
 - 5 * (4 * (3 * (2 * 1)))
 - 5 * (4 * (3 * 2))
 - 5 * (4 * 6)
 - 5 * 24
 - 120
 
递归归函数的优点是定义简单，逻辑清晰。理论上，所有的递归函数都可以写成循环的方式，但循环的逻辑不如递归清晰。

使用递归函数需要注意防止栈溢出。在计算机中，函数调用是通过栈（stack）这种数据结构实现的，每当进入一个函数调用，栈就会加一层栈帧，每当函数返回，栈就会减一层栈帧。由于栈的大小不是无限的，所以，递归调用的次数过多，会导致栈溢出。可以试试fact(1000)：

In [2]:
fact(10000)

RecursionError: maximum recursion depth exceeded in comparison

解决递归调用栈溢出的方法是通过尾递归优化，事实上尾递归和循环的效果是一样的，所以，把循环看成是一种特殊的尾递归函数也是可以的。

尾递归是指，在函数返回的时候，调用自身本身，并且，return语句不能包含表达式。这样，编译器或者解释器就可以把尾递归做优化，使递归本身无论调用多少次，都只占用一个栈帧，不会出现栈溢出的情况。

上面的fact(n)函数由于return n * fact(n - 1)引入了乘法表达式，所以就不是尾递归了。要改成尾递归方式，需要多一点代码，主要是要把每一步的乘积传入到递归函数中：

In [3]:
def fact(n):
    return fact_iter(n, 1)

def fact_iter(num, product):
    if num == 1:
        return product
    return fact_iter(num - 1, num * product)

可以看到，return fact_iter(num - 1, num * product)仅返回递归函数本身，num - 1和num * product在函数调用前就会被计算，不影响函数调用。
fact(5)对应的fact_iter(5, 1)的调用如下：

- fact_iter(5, 1)
- fact_iter(4, 5)
- fact_iter(3, 20)
- fact_iter(2, 60)
- fact_iter(1, 120)
- 120
尾递归调用时，如果做了优化，栈不会增长，因此，无论多少次调用也不会导致栈溢出。

遗憾的是，大多数编程语言没有针对尾递归做优化，Python解释器也没有做优化，所以，即使把上面的fact(n)函数改成尾递归方式，也会导致栈溢出。

In [5]:
fact(10000)

RecursionError: maximum recursion depth exceeded in comparison

### 小结
1. 使用递归函数的优点是逻辑简单清晰，缺点是过深的调用会导致栈溢出。
2. 针对尾递归优化的语言可以通过尾递归防止栈溢出。尾递归事实上和循环是等价的，没有循环语句的编程语言只能通过尾递归实现循环。
3. Python标准的解释器没有针对尾递归做优化，任何递归函数都存在栈溢出的问题。

### 练习
汉诺塔的移动可以用递归函数非常简单地实现。
请编写move(n, a, b, c)函数，它接收参数n，表示3个柱子A、B、C中第1个柱子A的盘子数量，然后打印出把所有盘子从A借助B移动到C的方法，例如：

In [14]:
def move(n, a, b, c):
    if n == 1:
        print(a, '-->', c)
    else:
        move(n-1,a, c, b)# 把 n-1 个盘子从 a 移动到 b 上
        move(1, a, b, c)# 把最底下的盘子从 a 移动到 c 上
        move(n-1, b, a, c)# 把 n-1 个盘子从 b 移动到 c 上
move(3, 'A', 'B', 'C')

A --> C
A --> B
C --> B
A --> C
B --> A
B --> C
A --> C


思路:如果要将n个盘子从a借助b移动到c，那么a中的最大的底盘移动到c时需要将n-1个盘子借助c移动到b上，然后再将最后的最大底盘移动到c，只不过是函数参数的顺序发生变化，抽象为函数即为：move(n-1, a, c, b)，move(1, a, b, c)，注意此时我们已经将a中的最大底盘移动到了c，b上面是n-1个盘子，现在的问题就回归到了b上面有n-1个盘子借助a移动到c的问题了，那么抽象为函数即为：move(n-1, b, a, c)。